-1 in A stride1s in GEMV

This is a strange bug that only seems to happen in Python 2, and only on some machines. It's been happening for a while, at least since the last release. I just never noticed it since it doesn't happen on my work computer.

Example error:

```
===================================================== FAILURES =====================================================
____________________________ test.nengo.networks.tests.test_actionselection.test_basic _____________________________

Simulator = <class 'nengo_ocl.tests.conftest.OclSimulator'>
plt = <nengo.utils.testing.Mock object at 0x7f57cd263f90>, seed = 1675708018

    def test_basic(Simulator, plt, seed):
        bg = nengo.networks.BasalGanglia(
            dimensions=5, net=nengo.Network(seed=seed))
        with bg:
            input = nengo.Node([0.8, 0.4, 0.4, 0.4, 0.4], label="input")
            nengo.Connection(input, bg.input, synapse=None)
            p = nengo.Probe(bg.output, synapse=0.01)

>       with Simulator(bg) as sim:

../nengo/nengo/networks/tests/test_actionselection.py:14: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
nengo_ocl/tests/conftest.py:13: in __init__
    super(OclSimulator, self).__init__(*args, context=ctx, **kwargs)
nengo_ocl/simulator.py:395: in __init__
    self._plan.extend(self.plan_op_group(op_type, op_list))
nengo_ocl/simulator.py:416: in plan_op_group
    return getattr(self, 'plan_' + op_type.__name__)(ops)
nengo_ocl/simulator.py:439: in plan_MultiDotInc
    tag='c-beta-%d' % len(constant_bs))
nengo_ocl/simulator.py:464: in _sig_gemv
    Y_in=Y_in, gamma=gamma, tag=tag)
nengo_ocl/clra_gemv.py:119: in __init__
    self.plans = self.choose_plans()
nengo_ocl/clra_gemv.py:1146: in choose_plans
    return block_impl(self, list(range(len(self.Y))))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

p = <nengo_ocl.clra_gemv.plan_block_gemv object at 0x7f57c208c550>, items = [0, 1, 2, 3, 4, 5, ...]

    def block_impl(p, items):

        if p.clra_alpha is not None:
            raise NotImplementedError()
        if p.clra_gamma is not None:
            raise NotImplementedError()
        if p.clra_beta is not None:
            raise NotImplementedError()
        if p.cl_alpha is not None:
            raise NotImplementedError()
        if p.cl_beta is not None:
            raise NotImplementedError()
        if p.cl_gamma is not None:
            raise NotImplementedError()
        if not all(s == 1 for s in p.A.stride1s):
>           raise NotImplementedError()
E           NotImplementedError

nengo_ocl/clra_gemv.py:847: NotImplementedError
============================================== pytest-warning summary ==============================================
WC1 None [pytest] section in setup.cfg files is deprecated, use [tool:pytest] instead.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=================================== 1 failed, 1 pytest-warnings in 2.14 seconds ====================================
```

--- a/nengo_ocl/__init__.py
+++ b/nengo_ocl/__init__.py
@@ -1,10 +1,20 @@
 import logging
 import sys
 
+import numpy as np
+import pyopencl as cl
+
 from nengo.utils.logging import log
 
 from .version import version as __version__
 from .simulator import Simulator
+
+numpy_relaxed_strides = np.ones((2, 1), order='C').flags.f_contiguous
+pyopencl_relaxed_strides = hasattr(cl.compyte.array, 'equal_strides')
+if numpy_relaxed_strides and not pyopencl_relaxed_strides:
+    raise ImportError("Numpy (v%s) uses relaxed strides, and PyOpenCL (v%s) does"
+                      " not support this. Please upgrade PyOpenCL." %
+                      (np.__version__, cl.version.VERSION_TEXT))
 
 # logging (default to no handler; use imported `log` fn to change this)
 try:

--- a/nengo_ocl/clraggedarray.py
+++ b/nengo_ocl/clraggedarray.py
@@ -10,6 +10,7 @@ from nengo.utils.compat import is_iterable, StringIO
 from pyopencl.array import Array, to_device
 
 from nengo_ocl.raggedarray import RaggedArray
+from nengo_ocl.utils import equal_strides
 
 # add 'ctype' property to Array (returned by 'to_device')
 Array.ctype = property(lambda self: cl.tools.dtype_to_ctype(self.dtype))
@@ -254,7 +255,8 @@ class CLRaggedArray(object):
                     array[...] = new_value
 
                 array.shape = clarray.shape  # reshape to avoid warning
-                assert array.strides == clarray.strides
+                assert equal_strides(
+                    array.strides, clarray.strides, clarray.shape)
                 clarray.set(array)
             else:
                 # discontiguous

--- a/nengo_ocl/simulator.py
+++ b/nengo_ocl/simulator.py
@@ -236,11 +236,13 @@ class ViewBuilder(object):
 
         assert obj.size and obj.ndim <= 2
         idx = self.sidx[obj.base]
+        shape0 = obj.shape[0] if obj.ndim > 0 else 1
+        shape1 = obj.shape[1] if obj.ndim > 1 else 1
         self.starts.append(self.rarray.starts[idx] + obj.elemoffset)
-        self.shape0s.append(obj.shape[0] if obj.ndim > 0 else 1)
-        self.shape1s.append(obj.shape[1] if obj.ndim > 1 else 1)
-        self.stride0s.append(obj.elemstrides[0] if obj.ndim > 0 else 1)
-        self.stride1s.append(obj.elemstrides[1] if obj.ndim > 1 else 1)
+        self.shape0s.append(shape0)
+        self.shape1s.append(shape1)
+        self.stride0s.append(obj.elemstrides[0] if shape0 > 1 else 1)
+        self.stride1s.append(obj.elemstrides[1] if shape1 > 1 else 1)
         self.names.append(getattr(obj, 'name', ''))
         self.sidx[obj] = len(self.sidx)
 

--- a/nengo_ocl/simulator.py
+++ b/nengo_ocl/simulator.py
@@ -605,7 +605,7 @@ class Simulator(nengo.Simulator):
     def _plan_fn_in_ocl(self, fn, tt, xx, yy, fn_name):
         signal_size = lambda sig: sig.size if sig is not None else None
         vector_dims = lambda shape, dim: len(shape) == 1 and shape[0] == dim
-        unit_stride = lambda es: len(es) == 1 and es[0] == 1
+        unit_stride = lambda s, es: len(es) == 1 and (s[0] == 1 or es[0] == 1)
 
         t_in = tt[0] is not None
         x_in = xx[0] is not None
@@ -619,10 +619,10 @@ class Simulator(nengo.Simulator):
         # for indexing errors)
         if x_in:
             assert all(vector_dims(x.shape, x_dim) for x in xx)
-            assert all(unit_stride(x.elemstrides) for x in xx)
+            assert all(unit_stride(x.shape, x.elemstrides) for x in xx)
 
         assert all(vector_dims(y.shape, y_dim) for y in yy)
-        assert all(unit_stride(y.elemstrides) for y in yy)
+        assert all(unit_stride(y.shape, y.elemstrides) for y in yy)
 
         # try to get OCL code
         in_dims = ([1] if t_in else []) + ([x_dim] if x_in else [])

--- a/nengo_ocl/tests/conftest.py
+++ b/nengo_ocl/tests/conftest.py
@@ -4,7 +4,7 @@ from nengo.utils.testing import allclose
 import pyopencl as cl
 import pytest
 
-from nengo.conftest import plt, rng  # noqa: F401
+from nengo.conftest import logger, plt, rng  # noqa: F401
 
 
 @pytest.fixture(scope="session")

--- a/nengo_ocl/tests/test_ast_conversion.py
+++ b/nengo_ocl/tests/test_ast_conversion.py
@@ -11,8 +11,6 @@ import nengo_ocl
 import nengo_ocl.ast_conversion as ast_conversion
 from nengo_ocl.ast_conversion import OCL_Function
 
-logger = logging.getLogger(__name__)
-
 
 @pytest.fixture(scope="session")
 def OclOnlySimulator(request, ctx):
@@ -197,14 +195,14 @@ def _test_conn(OclOnlySimulator, fn, size_in, dist_in=None, n=1):
 
     # compare output
     z = np.array([sim.data[p][-1] for p in probes])
-    assert np.allclose(z, y)
+    assert np.allclose(z, y, atol=2e-7)
 
 
 def test_sin_conn(OclOnlySimulator):
     _test_conn(OclOnlySimulator, np.sin, 1, n=10)
 
 
-def test_functions(OclOnlySimulator, n_points=10):
+def test_functions(OclOnlySimulator, logger, n_points=10):
     """Test the function maps in ast_converter.py"""
     # TODO: split this into one test per function using py.test utilities
 
@@ -279,7 +277,7 @@ def test_functions(OclOnlySimulator, n_points=10):
                         "see logger warnings for details")
 
 
-def test_vector_functions(OclOnlySimulator):
+def test_vector_functions(OclOnlySimulator, logger):
     d = 5
     boolean = [any, all, np.any, np.all]
     funcs = ast_conversion.vector_funcs.keys()

--- a/nengo_ocl/utils.py
+++ b/nengo_ocl/utils.py
@@ -18,6 +18,21 @@ def as_ascii(string):
         return string
 
 
+def equal_strides(strides1, strides2, shape):
+    """Check whether two arrays have equal strides.
+
+    Code from https://github.com/inducer/compyte
+    """
+    if len(strides1) != len(strides2) or len(strides2) != len(shape):
+        return False
+
+    for s, st1, st2 in zip(shape, strides1, strides2):
+        if s != 1 and st1 != st2:
+            return False
+
+    return True
+
+
 def get_closures(f):
     return OrderedDict(zip(
         f.__code__.co_freevars, (c.cell_contents for c in f.__closure__)))

